### PR TITLE
[BEAM-2748] accept SerializableFunction as UDF

### PR DESCRIPTION
**Pull request by XuMingmin**
R: @takidau @xumingming

### CODE REVIEW

**Comment by xumingming** [comment]
So now our UDF is a SerializableFunction, not a Clazz with a static `eval` method, right? Calcite supports optional arguments in UDF, will the new appoarch support this feature?

**Comment by XuMingmin** [comment]
The existing `BeamSqlUdf` is still there. Here it brings support of `SerializableFunction` as a second option. 
As discussed `SerializableFunction` can be used for one-in-one-out function, and you need to use `BeamSqlUdf` for cases with multiple parameters and/or optional parameter.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/BeamSqlUdfExpression.java [hunk 1]
```diff
@@ -78,6 +79,7 @@ private void reConstructMethod() {
       for (String pc : paraClassName) {
         paraClass.add(Class.forName(pc));
       }
+      udfIns = Class.forName(className).newInstance();
```
Comment (takidau) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/BeamSqlUdfExpression.java]: So for the static eval method version, this just returns the global static class instance? Or null? Or?

**Comment by takidau** [comment]
Merged, feel free to close after you answer my question above. Thanks!

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/BeamSqlUdfExpression.java (hunk 1 above)
Comment (XuMingmin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/BeamSqlUdfExpression.java]: It returns a new class instance every time. Let me include a logic to set it as null for static eval in another PR, although it doesn't take too much resource,

**Comment by XuMingmin** [comment]
close #3707, the left question is updated in #3710.

**Pull request closed without merging**